File: Day8Part1.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

class Day8Part1 {

    public static void main(String[] args) throws Exception {
        List<String> program = Files.lines(Paths.get("input-day8.txt")).collect(Collectors.toList());
        System.out.println(execute(program));
    }

    static int execute(List<String> program) {
        Set<Integer> executed = new HashSet<>();
        int accumulator = 0;
        int index = 0;
        while (true) {
            //System.out.println("index=" + index + " exec=" + executed);
            if (executed.contains(index)) return accumulator;
            executed.add(index);
            String[] tokens = program.get(index).split(" ");
            switch (tokens[0]) {
                case "acc":
                    accumulator += Integer.parseInt(tokens[1]);
                    index++;
                    break;
                case "jmp":
                    index += Integer.parseInt(tokens[1]);
                    break;
                case "nop":
                    index++;
                    break;
            }
        }
    }
}

File: day8part1.clj
(defn read-lines 
  [filename] 
  (vec (.split (slurp filename) "\n")))

(defn execute
  ([program] 
   (execute program 0 0 #{} []))
  ([program accum index indexes executed]
   ;;(println "acc" accum "indx" index indexes executed)
   (cond 
     (contains? indexes index)
     {:finished false :index index :accum accum :statements executed}

     (> index (count program))
     {:finished true :index index :accum accum :statements executed}

     :default
     (let [line (nth program index)
           tokens (.split line " ")
           op (first tokens)
           arg (Integer. (second tokens))
           executed (conj executed line)
           indexes (conj indexes index)]
       (condp = op
         "acc" (execute program (+ accum arg) (inc index) indexes executed)
         "jmp" (execute program accum (+ index arg) indexes executed)
         "nop" (execute program accum (inc index) indexes executed))))))

(println (execute (read-lines "input-day8.txt")))
  
